library/process/src/unixMain/kotlin/io/matthewnelson/kmp/process/internal/PlatformBuilder.kt

```diff
@@ -161,39 +166,126 @@ internal actual class PlatformBuilder private actual constructor() {
+                            // parent.
+                            IOException()
+                        }
                         else -> null
                     }
                 }
             } catch (_: IOException) {
-                // TODO: Handle error better
+                // [#, #, #, #, 1] (1: dup2 failure)
+                val b = ByteArray(5)
+                b[4] = 1
+                (err ?: EBADF).toBigEndian().copyInto(b)
+                try {
+                    StdioWriter(pipe).write(b)
+                } finally {
+                    close(pipe.fdWrite)
+                }
                 _exit(1)
             }
 
-            memScoped {
+            val errno = memScoped {
                 val argv = args.toArgv(program = program, scope = this)
                 val envp = env.toEnvp(scope = this)
 
                 execve(program.path, argv, envp)
-                // TODO: Handle error better
-                _exit(errno)
+
+                // exec failed to replace child process with program
+                errno
             }
+
+            // [#, #, #, #, 2] (2: execve failure)
+            val b = ByteArray(5)
+            b[4] = 2
+            errno.toBigEndian().copyInto(b)
+            try {
+                StdioWriter(pipe).write(b)
+            } finally {
+                close(pipe.fdWrite)
+            }
+            _exit(1)
         }
 
-        return NativeProcess(
+        // Parent process
+        close(pipe.fdWrite)
+
+        val p = NativeProcess(
             pid,
             handle,
             command,
             args,
             env,
             destroy,
         )
+
+        // Below is "sort of" like vfork on Linux, but
+        // with error validation and cleanup on our end.
+        val b = ByteArray(5)
+
+        val read = try {
+            StdioReader(pipe).read(b)
+        } catch (e: IOException) {
+            p.destroy()
+            throw IOException("CLOEXEC pipe read failure", e)
+        } finally {
+            close(pipe.fdRead)
+        }
+
+        var err: IOException? = null
+
+        when (read) {
+            0 -> { /* execve successful */ }
+            b.size -> {
+                val type = when (b[4].toInt()) {
+                    1 -> "dup2"
+                    2 -> "exec"
+                    else -> null
+                }
+
+                err = if (type == null) {
+                    IOException("CLOEXEC pipe validation check failure")
+                } else {
+                    val errno = BigEndian(b[0], b[1], b[2], b[3]).toInt()
+                    val msg = strerror(errno)?.toKString() ?: "errno: $errno"
+                    IOException("Child process $type failure. $msg")
+                }
+            }
+            -1 -> err = errnoToIOException(errno)
+            else -> err = IOException("invalid read on CLOEXEC pipe")
+        }
+
+        err?.let { ex ->
+            p.destroy()
+            throw ex
+        }
```

Comment:
This could be cleaner w/o need to use `var`.

also, `-1` will never be returned as `read` throws exception for that.

Reply:
Resolved in commit 4c9207b